src/System.Private.CoreLib/src/System/Runtime/CompilerServices/ConditionalWeakTable.cs

```diff
@@ -424,6 +433,22 @@ internal bool Remove(TKey key)
             }
 
             //----------------------------------------------------------------------------------------
+            // Removes all entries from the table.
+            //
+            // Precondition:
+            //      Must hold _lock.
+            //----------------------------------------------------------------------------------------
+            internal void Clear()
+            {
+                // Remove all handles (as in Remove, just by setting the hashcode and leaving
+                // actual removal up to the finalizer).
+                for (int i = 0; i < _firstFreeEntry; i++)
```

Comment:
> Was the same problem in the original CoreCLR implementation

The original CoreCLR implementation doesn't have to worry about lock-free reads, so it frees all of the handles synchronously in Clear.

> We should also set the primary to null or something

Can we do that while a reader may be using the dependency handle? i.e. does the handle itself have appropriate synchronization to support reading-while-mutating?

I'd based this Clear implementation on the existing Remove implementation, which also just sets the hashcode to -1.  Presumably whatever we do in Clear should also be done in Remove.